### PR TITLE
DP-342 Test Datasource Upon Service Creation

### PR DESCRIPTION
**Pull request by tomonorman**

### CODE REVIEW

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 1]
```diff
@@ -612,11 +635,39 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                                        // Update our message Options. A succesful save may still result in an unsuccesful error so we may
+                                        // need to reassign the type
+                                        messageOptions.type = result.type;
+                                        messageOptions.message = messageOptions.message + ' ' + result.message;
+                                        // Send out the notification to the user.
+                                        messageOptions.type === 'success' ? dfNotify.success(messageOptions) : dfNotify.error(messageOptions);
+
+                                        if (messageOptions.type === 'success') {
+                                            scope.closeEditor();
+                                        }
```
Comment (yaroslavmo) [inline app/admin_components/adf-services/dreamfactory-services.js:657]: Lines 664-666 are identical to these.

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 2]
```diff
@@ -597,9 +622,7 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
 
                                 };
```
Comment (yaroslavmo) [inline app/admin_components/adf-services/dreamfactory-services.js:623]: Why don't just remove this first `then` as you still added a new one with testServiceConnection after it?

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 3]
```diff
@@ -667,14 +718,34 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
                                     type: 'error',
                                     provider: 'dreamfactory',
                                     message: reject
-
+
```
Comment (yaroslavmo) [inline app/admin_components/adf-services/dreamfactory-services.js:721]: remove space

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 4]
```diff
@@ -667,14 +718,34 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                                        // need to reassign the type
+                                        messageOptions.type = result.type;
+                                        messageOptions.message = messageOptions.message + ' ' + result.message;
+                                        // Send out the notification to the user.
+                                        messageOptions.type === 'success' ? dfNotify.success(messageOptions) : dfNotify.error(messageOptions);
+                                    })
+                                } else {
+                                    // if not a database connector just send out the notification of succesful save or not.
+                                    messageOptions.type === 'success' ? dfNotify.success(messageOptions) : dfNotify.error(messageOptions);
+                                }
```
Comment (yaroslavmo) [inline app/admin_components/adf-services/dreamfactory-services.js:745]: the code is identical to the save request?  You should have moved it to a function.

**Review by yaroslavmo — CHANGES_REQUESTED** [review]

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 5]
```diff
@@ -571,13 +571,38 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                            return  {
+                                type: 'success',
+                                message:'Test connection succeeded.'
+                            };
+
+                        }, function(reject) {
+
+                            return {
+                                type: 'error',
+                                message: 'However Test Connection Failed: ' + reject.data.error.message
```
[the remark this answers is not in the document]
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:589]: DOne

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 6]
```diff
@@ -571,13 +571,38 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                    scope.isServiceTypeDatabase = function () {
+                        return scope.selectedSchema.group == 'Database';
+                    };
+
                     scope.saveService = function () {
 
                         // merge data from UI into current edit record
                         scope.prepareServiceData();
-
+
```
[the remark this answers is not in the document]
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:603]: done

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 7]
```diff
@@ -571,13 +571,38 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
                     scope.saveService = function () {
 
                         // merge data from UI into current edit record
                         scope.prepareServiceData();
-
+                        
                         var data = normalizeKeyValuePairs();
-
+
```
[the remark this answers is not in the document]
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:605]: sorry

File: app/admin_components/adf-services/dreamfactory-services.js (hunk 2 above)
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:623]: Sorry, I'm not following this one -> this github preview is just showing some blank lines.
Do you mean the `then`  on line 614, 640, and 647? I needed to chain these thens as the calls to the backend are all asynchrous so needed to chain the promises to stop one returning before a different one (i.e keep the order the same). 

So each function is returning a result, which its passing to the next one. 

Or am i up the wrong tree?

File: app/admin_components/adf-services/dreamfactory-services.js (hunk 1 above)
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:657]: moved the stuff above it into a function, and now this conditional is called after the function completes

File: app/admin_components/adf-services/dreamfactory-services.js (hunk 4 above)
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:745]: whoopsy, done

File: app/admin_components/adf-services/dreamfactory-services.js (hunk 3 above)
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:721]: done

File: app/admin_components/adf-services/dreamfactory-services.js (hunk 2 above)
Comment (yaroslavmo) [reply app/admin_components/adf-services/dreamfactory-services.js:643]: Yeah sorry got lost in the callback hell

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 8]
```diff
@@ -571,6 +571,51 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                            };
+
+                        });
+                    };
+
+                    scope.isServiceTypeDatabase = function () {
+                        return scope.selectedSchema.group == 'Database';
+                    };
+
+                    var displayServiceConnectionNotification = function (messageOptions) {
```
Comment (yaroslavmo) [inline app/admin_components/adf-services/dreamfactory-services.js:599]: `displayServiceConnectionNotification` is a bad name for this function as it is not only displaying but actually doing the request.

`testServiceConnection` is way better but I see that you used it inside of the function.
I suggest renaming your current testServiceConnection to something like `testServiceSchema` and call the main function `testServiceConnection`

**Review by yaroslavmo — CHANGES_REQUESTED** [review]

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 9]
```diff
@@ -571,6 +571,51 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                            return {
+                                type: 'error',
+                                message: 'Test Connection Failed: ' + reject.data.error.message
+                            };
+
+                        });
+                    };
+
+                    scope.isServiceTypeDatabase = function () {
+                        return scope.selectedSchema.group == 'Database';
```
[the remark this answers is not in the document]
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:596]: oops, sorry, done

File: app/admin_components/adf-services/dreamfactory-services.js (hunk 8 above)
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:599]: Good call, done

File: app/admin_components/adf-services/dreamfactory-services.js [hunk 10]
```diff
@@ -571,6 +571,51 @@ angular.module('dfServices', ['ngRoute', 'dfUtility', 'dfApplication'])
...
+                        if (scope.isServiceTypeDatabase()) {
+                            // Fire the test connection function, and wait for the promise to resolve.
+                            var testResults = scope.testServiceConnection();
+                                testResults.then(function (result) {
+                                    // Update our message Options. A succesful save may still result in an unsuccesful error so we may
+                                    // need to reassign the type
+                                    messageOptions.type = result.type;
+                                    messageOptions.message = messageOptions.message + ' ' + result.message;
+                                    // Send out the notification to the user.
+                                    messageOptions.type === 'success' ? dfNotify.success(messageOptions) : dfNotify.error(messageOptions);
```
[the remark this answers is not in the document]
Comment (tomonorman) [reply app/admin_components/adf-services/dreamfactory-services.js:611]: ah yeah. Done

**Comment by yaroslavmo** [comment]
@tomonorman I found another bug. Here is the flow:
When I am trying to create a service with the same namespace that already exists in the system I expect an error saying that there is a duplicate entry:
`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'test' for key` 
But this error is overwritten with an error from your test request.

The only way how we can fix that is to create a button "test connection" so users would click on it instead of doing this test automatically. This way we would not slip into dreamfactory flow and overwrite any important notifications.

**Comment by tomonorman** [comment]
HI @yaroslavmo 
My latest commit has a check whether the initial save was succesful or not. If it wasnt, then it will ignore the testing logic and just output the error (as it used to do)

This should get rid of any issues with the test overwriting a failed save?

**Review by yaroslavmo — APPROVED** [review]

**Pull request merged**